drivers/bluetooth/hci/rpmsg_nrf53.c

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ipm.h>
+
+#include <openamp/open_amp.h>
+#include <metal/sys.h>
+#include <metal/device.h>
+#include <metal/alloc.h>
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HCI_DRIVER)
+#define LOG_MODULE_NAME bt_hci_driver_nrf53
+#include "common/log.h"
+
+void bt_rpmsg_rx(u8_t *data, size_t len);
+
+static K_SEM_DEFINE(sync_sem, 0, 1);
+
+static struct device *ipm_tx_handle;
+static struct device *ipm_rx_handle;
+
+/* Configuration defines */
+
+#define SHM_START_ADDR      (DT_IPC_SHM_BASE_ADDRESS + 0x400)
+#define SHM_SIZE            0x7c00
+#define SHM_DEVICE_NAME     "sram0.shm"
+
+#define VRING_COUNT         2
+#define VRING_TX_ADDRESS    (SHM_START_ADDR + SHM_SIZE - 0x400)
+#define VRING_RX_ADDRESS    (VRING_TX_ADDRESS - 0x400)
+#define VRING_ALIGNMENT     4
+#define VRING_SIZE          16
+
+#define VDEV_STATUS_ADDR    DT_IPC_SHM_BASE_ADDRESS
+
+/* End of configuration defines */
+
+static metal_phys_addr_t shm_physmap[] = { SHM_START_ADDR };
+static struct metal_device shm_device = {
+	.name = SHM_DEVICE_NAME,
+	.bus = NULL,
+	.num_regions = 1,
+	.regions = {
+		{
+			.virt       = (void *) SHM_START_ADDR,
+			.physmap    = shm_physmap,
+			.size       = SHM_SIZE,
+			.page_shift = 0xffffffff,
+			.page_mask  = 0xffffffff,
+			.mem_flags  = 0,
+			.ops        = { NULL },
+		},
+	},
+	.node = { NULL },
+	.irq_num = 0,
+	.irq_info = NULL
+};
+
+static struct virtqueue *vq[2];
+static struct rpmsg_endpoint ep;
+
+static unsigned char virtio_get_status(struct virtio_device *vdev)
+{
+	return VIRTIO_CONFIG_STATUS_DRIVER_OK;
+}
+
+static void virtio_set_status(struct virtio_device *vdev, unsigned char status)
+{
+	sys_write8(status, VDEV_STATUS_ADDR);
+}
+
+static u32_t virtio_get_features(struct virtio_device *vdev)
+{
+	return BIT(VIRTIO_RPMSG_F_NS);
+}
+
+static void virtio_set_features(struct virtio_device *vdev, u32_t features)
+{
+	/* No need for implementation */
+}
+
+static void virtio_notify(struct virtqueue *vq)
+{
+	int status;
+
+	status = ipm_send(ipm_tx_handle, 0, 0, NULL, 0);
+	if (status != 0) {
+		BT_ERR("ipm_send failed to notify: %d", status);
+	}
+}
+
+const struct virtio_dispatch dispatch = {
+	.get_status = virtio_get_status,
+	.set_status = virtio_set_status,
+	.get_features = virtio_get_features,
+	.set_features = virtio_set_features,
+	.notify = virtio_notify,
+};
+
+static void ipm_callback(void *context, u32_t id, volatile void *data)
+{
+	BT_DBG("Got callback of id %u", id);
+	virtqueue_notification(vq[0]);
```

Comment:
Same issue,  processing in interrupt context

Reply:
I tried to switch context for ```virtqueue_notification``` from ISR to the workqueue context, but oddly enough this triggers MPU fault.

There is a flash write attempt that crashes the application. It occurs inside ```sys_dlist_remove``` in the ```k_sys_work_q``` context. Do you see any potential scenario in which the MPU fault could be caused by OpenAMP library (e.g. pointer corruption)?

Reply:
However, using a dedicated thread to process ```virtqueue_notification``` seems to work.

Reply:
if you are sending  a message you can take rdev->lock (rpmsg_virtio_send_offchannel_raw). And ipm_callback is called you will treat received messages , try to take rdev->lock a second time(rpmsg_virtio_rx_callback) => deadlock 
yes using a dedicated thread is a good solution

Reply:
Added a dedicated thread for ```virtqueue_notification```